nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilderFacetImpl.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Sonatype Nexus (TM) Open Source Version
+ * Copyright (c) 2017-present Sonatype, Inc.
+ * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.r.internal;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.sonatype.nexus.common.event.EventAware.Asynchronous;
+import org.sonatype.nexus.common.event.EventManager;
+import org.sonatype.nexus.common.stateguard.Guarded;
+import org.sonatype.nexus.repository.FacetSupport;
+import org.sonatype.nexus.repository.r.RHostedFacet;
+import org.sonatype.nexus.repository.r.RPackagesBuilderFacet;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetCreatedEvent;
+import org.sonatype.nexus.repository.storage.AssetDeletedEvent;
+import org.sonatype.nexus.repository.storage.AssetEvent;
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.STARTED;
+import static org.sonatype.nexus.repository.r.internal.AssetKind.ARCHIVE;
+import static org.sonatype.nexus.repository.r.internal.RPathUtils.getBasePath;
+
+/**
+ * Implementation of {@link RPackagesBuilderFacet} targeted for use with hosted repositories. Uses event dispatching to
+ * ensure that we do not have race conditions when processing and rebuilding metadata, and also imposes a waiting period
+ * to batch metadata updates.
+ */
+@Named
```

Comment:
Please add 'since 1.next' value

Reply:
done in fe40919080a00ea9162d0af9be02c414af338e27